src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -288,43 +327,100 @@ private async Task SaveAsync()
                 this.pendingTipHashHeight = null;
             }
 
-            this.CheckItemsAreInSequence(pendingBatch.Keys.ToList());
+            this.CheckItemsAreInConsecutiveSequence(pendingBatch.Keys.ToList());
 
+            // Save the items to disk.
             using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
             {
                 // Save the items to disk.
-                await this.provenBlockHeaderRepository.PutAsync(pendingBatch.Values.ToList(), hashHeight);
+                await this.provenBlockHeaderRepository.PutAsync(pendingBatch.Values.ToList(), hashHeight).ConfigureAwait(false);
+
+                this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+            }
+        }
+
+        /// <summary>
+        /// Will set the <see cref="IProvenBlockHeaderStore"/> tip to the most recent <see cref="ProvenBlockHeader"/> that exists in the <see cref="IProvenBlockHeaderRepository"/> and/or <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task<ChainedHeader> RecoverStoreTipAsync()
+        {
+            int tipHeight = this.provenBlockHeaderRepository.TipHashHeight.Height;
 
-                this.TipHashHeight = hashHeight;
+            ProvenBlockHeader latestHeader = await this.provenBlockHeaderRepository.GetAsync(tipHeight);
+
+            if ((tipHeight > 0) && (latestHeader == null))
+            {
+                // Happens when the proven header store is corrupt.
+                throw new ProvenBlockHeaderException("Proven block header store failed to recover.");
+            }
+
+            uint256 latestBlockHash = this.provenBlockHeaderRepository.TipHashHeight.Hash;
+
+            while (this.chain.GetBlock(tipHeight) == null)
+            {
+                if (latestHeader.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    latestBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                latestHeader = await this.provenBlockHeaderRepository.GetAsync(tipHeight--).ConfigureAwait(false);
```

Comment:
no need to get it from repo every time. You need to get it only when we have it in concurrent chain. 

Basically we are looking for last common header between chain and repo

Reply:
I've applied the change, but would like you to double check I've covered all cases.

Reply:
Changed the logic in RecoverStoreTipAsync() quite a bit. Please review.